waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,311 @@
+    debug "sync session ended",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId, error
+    return err("Sync request error: " & error)
+
+  return ok((hashes, peer))
+
+proc sync*(
+    self: WakuSync, peer: RemotePeerInfo
+): Future[Result[seq[WakuMessageHash], string]] {.async, gcsafe.} =
+  let connOpt = await self.peerManager.dialPeer(peer, WakuSyncCodec)
+  let conn: Connection = connOpt.valueOr:
+    return err("Cannot establish sync connection")
+
+  let hashes: seq[WakuMessageHash] = (await self.request(conn)).valueOr:
+    debug "sync session ended",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId, error
+
+    return err("Sync request error: " & error)
+
+  return ok(hashes)
+
+proc handleLoop(
+    self: WakuSync, conn: Connection
+): Future[Result[seq[WakuMessageHash], string]] {.async.} =
+  let (start, `end`) = calculateRange(self.relayJitter)
+
+  let initialized =
+    ?serverInitialize(self.storage, conn, self.maxFrameSize, start, `end`)
+
+  var sent = initialized
+
+  while true:
+    let received = ?await sent.listenBack()
+
+    trace "sync payload received",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = received.payload
+
+    let reconciled = (?received.serverReconcile()).valueOr:
+      let completed = error # Result[Reconciled, Completed]
+
+      let hashes = await completed.serverTerminate()
+
+      return ok(hashes)
+
+    sent = ?await reconciled.send()
+
+    trace "sync payload sent",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = reconciled.payload
+
+proc initProtocolHandler(self: WakuSync) =
+  proc handle(conn: Connection, proto: string) {.async, gcsafe, closure.} =
+    debug "sync session requested",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+    let hashes = (await self.handleLoop(conn)).valueOr:
+      debug "sync session ended",
+        server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId, error
+
+      #TODO send error code and desc to client
+      return
+
+    if hashes.len > 0 and self.transferCallBack.isSome():
+      let callback = self.transferCallBack.get()
+
+      (await callback(hashes, conn.peerId)).isOkOr:
+        debug "transfer callback failed", error
+
+    debug "sync session ended gracefully",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+  self.handler = handle
+  self.codec = WakuSyncCodec
+
+proc new*(
+    T: type WakuSync,
+    peerManager: PeerManager,
+    maxFrameSize: int = DefaultMaxFrameSize,
+    syncInterval: timer.Duration = DefaultSyncInterval,
+    relayJitter: Duration = DefaultGossipSubJitter,
+    pruneCB: Option[PruneCallBack] = none(PruneCallback),
+    transferCB: Option[TransferCallback] = none(TransferCallback),
+): T =
+  let storage = Storage.new().valueOr:
+    debug "storage creation failed"
```

Comment:
Maybe make this an error log...I don't remember why we couldn't log error returned here.
I was facing some weird compilation issue.